fw/http_limits.c

```diff
@@ -639,72 +639,55 @@ frang_http_host_check(const TfwHttpReq *req, FrangAcc *ra)
 	unsigned short port;
 	unsigned short real_port;
 	TfwStr prim_trim = { 0 }, prim_name = { 0 }, /* primary source */
-	       sec_trim = { 0 },  sec_name = { 0 },  /* secondary source */
-	       fwd_trim = { 0 },  fwd_name = { 0 },
-	       *val_trim = NULL;  /* effective source */
+	       fwd_trim = { 0 },  fwd_name = { 0 };
 
 	BUG_ON(!req);
 	BUG_ON(!req->h_tbl);
 
 	switch (req->version) {
-	/*
-	 * In h2 protocol the Host: header is not required and :authority
-	 * pseudo-header should be used instead. But this is "SHOULD"
-	 * requirement and a request may has missing :authority and existing
-	 * Host: header. Block request if none defined or their values are
-	 * not identical.
-	 */
-	case TFW_HTTP_VER_20:
-		frang_get_host_header(req, TFW_HTTP_HDR_H2_AUTHORITY,
-				      &prim_trim, &prim_name);
-		frang_get_host_header(req, TFW_HTTP_HDR_HOST,
-				      &sec_trim, &sec_name);
-		val_trim = &prim_trim;
-		if (unlikely(TFW_STR_EMPTY(val_trim))) {
-			val_trim = &sec_trim;
-			if (unlikely(TFW_STR_EMPTY(val_trim))) {
-				frang_msg("Request authority is unknown",
-					  &FRANG_ACC2CLI(ra)->addr, "\n");
-				return TFW_BLOCK;
-			}
-		}
-		else if (frang_assert_host_header(&prim_trim, &sec_trim)) {
-			frang_msg("Request authority differs between headers",
-				  &FRANG_ACC2CLI(ra)->addr, "\n");
-			return TFW_BLOCK;
-		}
-		else if (frang_get_host_forwarded(req, &fwd_trim, &fwd_name)
-			 && frang_assert_host_header(&prim_trim, &fwd_trim)) {
-			frang_msg("Request authority differs from forwarded",
+	case TFW_HTTP_VER_20: {
+		TfwStr authority, host;
+		__h2_msg_hdr_val(&req->h_tbl->tbl[TFW_HTTP_HDR_H2_AUTHORITY],
+		                 &authority);
+		__h2_msg_hdr_val(&req->h_tbl->tbl[TFW_HTTP_HDR_HOST], &host);
+		/* https://datatracker.ietf.org/doc/html/rfc9113#section-8.3.1
+		 * A server SHOULD treat a request as malformed if it contains
+		 * a Host header field that identifies an entity that differs
+		 * from the entity in the ":authority" pseudo-header field.
+                 *
+                 * Note: check_authority_correctness() already ensures that
+                 * at least one of :authority or Host headers are not empty.
+		 */
+		if (TFW_STR_EMPTY(&authority) || TFW_STR_EMPTY(&host)
+                    || tfw_strcmp(&authority, &host) != 0)
+			frang_msg("Request :authority differs from Host",
 				  &FRANG_ACC2CLI(ra)->addr, "\n");
-			return TFW_BLOCK;
-		}
-		break;
-	/*
-	 * In http/1.1 host header and authority defined in URI must always be
-	 * identical.
-	 */
+                        return TFW_BLOCK;
```

Comment:
`if` has no braces, so we never reach `fallthrough`. I think it was a bad idea to missallign braces just to make `authority` and `host` variables local.